src/wt.ts

```diff
@@ -0,0 +1,357 @@
+import QS from "qs";
+import Cookie from "js-cookie";
+import EventEmitter from "events";
+import { debounce, isFunction, omitBy, isNil, uuid } from "./utils";
+
+const DEBOUNCE_MIN_DEFAULT = 500;
+const DEBOUNCE_MAX_DEFAULT = 1500;
+
+const DEFAULT_STRINGIFY_OPTIONS = {
+  arrayFormat: "brackets",
+  skipNulls: true,
+  encode: true,
+} as const;
+
+// global constants
+const BATCH_MAX = 100;
+const EXPIRES_IN_DAYS = 7300; // Expires in 20 years
+
+function resolveMethod<V>(val: V | ((...a: any[]) => V), ...args: any[]) {
+  return isFunction(val) ? val(...args) : val;
+}
+
+export const SEND_STARTED = "send:started";
+export const SEND_COMPLETED = "send:completed";
+export const QUEUE_COMPLETED = "queue:completed";
+export const QUEUE_CONTINUED = "queue:continued";
+
+export const VISITOR_TOKEN_KEY = "wt_visitor_token";
+export const PAGE_UUID_KEY = "wt_page_uuid";
+
+function retrieveFromCookie(key: string, config: Cookie.CookieAttributes = {}) {
+  let token = Cookie.get(key);
+  if (!token) {
+    token = uuid();
+    Cookie.set(key, token, config);
+  }
+  return token;
+}
+
+const retrieveFromQueryString = (search: string) => {
+  const qs = QS.parse(search, { ignoreQueryPrefix: true });
+  return qs.wvt as string | undefined;
+};
+
+const retrieveVisitorToken = (
+  config: Cookie.CookieAttributes = {},
+  search: string
+) =>
+  retrieveFromQueryString(search) ||
+  retrieveFromCookie(VISITOR_TOKEN_KEY, config);
+
+const retrievePageUUIDToken = () => retrieveFromCookie(PAGE_UUID_KEY);
+
+type WTConfig = {
```

Comment:
can we namespace these like WT.Config?

Reply:
this is actually pretty awkward to do with autogenerated types so opting to defer for now. i'd like to figure out how to handle this better, but would prefer to get this out (since we largely don't need to import these types, impact should be minimal).